### PR TITLE
SoundWire: better error handling in deferred transfers

### PR DESCRIPTION
**Pull request by plbossart**
Generic cleanups while fighting with errors during 192 kHz setup (https://github.com/thesofproject/linux/pull/4057)

### CODE REVIEW

**Review by bardliao — DISMISSED** [review]

**Review by RanderWang — DISMISSED** [review]
LGTM

File: drivers/soundwire/stream.c [hunk 1]
```diff
@@ -1389,7 +1389,7 @@ static int _sdw_prepare_stream(struct sdw_stream_runtime *stream,
 
 	ret = do_bank_switch(stream);
 	if (ret < 0) {
-		dev_err(bus->dev, "Bank switch failed: %d\n", ret);
+		dev_err(bus->dev, "do_bank_switch failed: %d\n", ret);
```
Comment (ranj063) [inline drivers/soundwire/stream.c:1392]: @plbossart is this even if you have the +pmf flag set for the module?

File: drivers/soundwire/bus.c [hunk 2]
```diff
@@ -237,7 +237,7 @@ static inline int do_transfer_defer(struct sdw_bus *bus,
 	init_completion(&defer->complete);
 
 	for (i = 0; i <= retry; i++) {
-		resp = bus->ops->xfer_msg_defer(bus, msg, defer);
```
Comment (ranj063) [inline drivers/soundwire/bus.c:240]: @plbossart it seems like even the msg arg is redundant isnt it? You already set the defer->msg = msg just a couple of lines above

File: drivers/soundwire/stream.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/stream.c:1392]: +pmf is for dev_dbg, for dev_err we have unfortunately no means to add more qualifiers.

File: drivers/soundwire/bus.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/bus.c:240]: that's removed in the next patch

Comment (ranj063) [reply drivers/soundwire/bus.c:240]: next patch? you mean you have more coming? this is the 4th patch in this series?

Comment (plbossart) [reply drivers/soundwire/bus.c:240]: just look at the patch, it's removed in the 4th....

Comment (plbossart) [reply drivers/soundwire/bus.c:240]: https://github.com/thesofproject/linux/pull/4070/commits/d44559361240e29acdd6f49ba46c6acab939a993#diff-5bcb345af1e70754ffdfe5c3189ab1e15f6c0f0ca53c6e258083562af1df4f29L240

Comment (ranj063) [reply drivers/soundwire/bus.c:240]: I'm sorry @plbossart. I'm not sure im following. Even after the 4th patch, "msg" is still an argument for the xfer_msg_defer op. What am I missing?

Comment (plbossart) [reply drivers/soundwire/bus.c:240]: oh sorry, I thought you talked about the defer argument....

**Review by ranj063 — DISMISSED** [review]
looks good @plbossart. Thanks!

File: drivers/soundwire/bus.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/bus.c:240]: done in new 5th patch

**Review by ranj063 — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

File: drivers/soundwire/cadence_master.c [hunk 3]
```diff
@@ -721,10 +721,11 @@ cdns_xfer_msg(struct sdw_bus *bus, struct sdw_msg *msg)
 EXPORT_SYMBOL(cdns_xfer_msg);
 
 enum sdw_command_response
-cdns_xfer_msg_defer(struct sdw_bus *bus,
-		    struct sdw_msg *msg, struct sdw_defer *defer)
+cdns_xfer_msg_defer(struct sdw_bus *bus)
 {
 	struct sdw_cdns *cdns = bus_to_cdns(bus);
+	struct sdw_defer *defer = &bus->defer_msg;
```
Comment (RanderWang) [inline drivers/soundwire/cadence_master.c:727]: no reference of variable defer in function now except  defer->msg,  change the following line to "struct sdw_msg *msg = bus->defer_msg.msg;" ?

File: drivers/soundwire/cadence_master.c [hunk 4]
```diff
@@ -881,13 +879,15 @@ irqreturn_t sdw_cdns_irq(int irq, void *dev_id)
 		return IRQ_NONE;
 
 	if (int_status & CDNS_MCP_INT_RX_WL) {
+		struct sdw_bus *bus = &cdns->bus;
+		struct sdw_defer *defer = &bus->defer_msg;
+
```
Comment (RanderWang) [inline drivers/soundwire/cadence_master.c:884]: we don't need define variable bus here,  use "struct sdw_defer *defer = cdns->bus.defer_msg;" ?

Comment (plbossart) [reply drivers/soundwire/cadence_master.c:884]: I like keeping indirections simple enough, and specifically make it clear that the defer stuff is a BUS structure.

**Pull request merged**